memcached.c

```diff
@@ -4181,6 +4263,740 @@ static inline void process_get_command(conn *c, token_t *tokens, size_t ntokens,
+                        won_token = true;
+                    }
+                    rtokens++; // always consume the token.
+                    break;
+                case 'R':
+                    // If we haven't autovivified and supplied token is less
+                    // than current TTL, mark a win.
+                    if ((it->it_flags & ITEM_TOKEN_SENT) == 0
+                            && !item_created
+                            && it->exptime != 0) {
+                        if (!safe_strtol(tokens[rtokens].value, &exptime_int)) {
+                            errstr = "bad tokens in command line format";
+                            goto error;
+                        }
+
+                        if (it->exptime - current_time < exptime_int) {
+                            won_token = true;
+                        }
+                    }
+                    rtokens++;
+                    break;
+                case 's':
+                    *p = ' ';
+                    p = itoa_u32(it->nbytes-2, p+1);
+                    break;
+                case 't':
+                    // TODO: ensure this is correct for autoviv case.
+                    // or, I guess users can put N before t?
+                    // TTL remaining as of this request.
+                    // needs to be relative because server clocks may not be in sync.
+                    *p = ' ';
+                    if (it->exptime == 0) {
+                        *(p+1) = '-';
+                        *(p+2) = '1';
+                        p += 3;
+                    } else {
+                        p = itoa_u32(it->exptime - current_time, p+1);
+                    }
+                    break;
+                case 'c':
+                    *p = ' ';
+                    p = itoa_u64(ITEM_get_cas(it), p+1);
+                    break;
+                case 'v':
+                    of.value = 1;
+                    break;
+                case 'f':
+                    *p = ' ';
+                    if (FLAGS_SIZE(it) == 0) {
+                        *(p+1) = '0';
+                        p += 2;
+                    } else {
+                        p = itoa_u32(*((uint32_t *) ITEM_suffix(it)), p+1);
+                    }
+                    break;
+                case 'l':
+                    // FIXME: this will be _now_ unless paired with 'u'. allow get command to pass back
+                    // pre-existing last-access time?
+                    *p = ' ';
+                    p = itoa_u32(current_time - it->time, p+1);
+                    break;
+                case 'h':
+                    *p = ' ';
+                    if (it->it_flags & ITEM_FETCHED) {
+                        *(p+1) = '1';
+                    } else {
+                        *(p+1) = '0';
+                    }
+                    break;
+                default:
+                    fprintf(stderr, "Unknown option: %c\n", opts[i]);
```

Comment:
What about `u` ? Or you didn't implement it until finding a good way to test it?

Reply:
Shouldn't we send an error message to the client for invalid options instead of printing to the `stderr` (probably based on the verbosity level)

Reply:
re: u, there's just no way to "fetch the item with a bump and get the last_access time before the bump" in the item API. Need to make that happen first. It's just a TODO.

re: printf. yes, that's a development/debug line right now. I wrote that before out_errstr existed.

Reply:
oh I misread you; the 'u' is handled in the preparser above. some common and idempotent flags are done up there in the common code and passed back with the of bits. 'u' already works.

Reply:
I'm a liar; 'u' wasn't properly implemented yet. it is now.